rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,447 @@
+  """
+
+  @abc.abstractmethod
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[str]:
+    """Define compressible weights for each layer.
+
+    Args:
+       original_layer: tf.keras.layers.Layer representing a layer from the
+       original model.
+
+    Returns:
+       List of attribute names as string representing list of compressible
+       weights for the given layer. (e.g. return value ['kernel'] means
+       layer.kernel is compressible.)
+    """
+
+  @abc.abstractmethod
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Initialize training weights for the training model. It calls the `add_training_weight` method several times to add training weights.
+
+    Args:
+      pretrained_weight: tf.Tensor of a pretrained weight of a layer that will
+        be compressed eventually.
+    """
+
+  def add_training_weight(
+      self, *args, **kwargs):
+    """Add training weight for the training model. This method is called from `init_training_weights`.
+
+    Args:
+      *args, **kwargs: args and kwargs for training_model.add_weight.
+    """
+
+  @abc.abstractmethod
+  def project_training_weights(self, *training_weights: tf.Tensor) -> tf.Tensor:
+    """Define a piece of the forward pass during training, which operates on a single compressible weight.
+    The default throws an error when training occurs.
+
+    Args:
+       *training_weights: tf.Tensors representing any variables used during
+         training, for a single compressible weight, in the order returned in
+         `init_training_weights`.
+
+    Returns:
+       tf.Tensor to set the compressible weight to.
+    """
+
+  def update_training_weight(self, index: integer, tensor: tf.Tensor):
+    """Update a training weight on an index to a given tensor value.
+
+    This method is for the case that training weight should update to specific
+    value not from the model optimizer. It'll throws an error if it can't
+    find the training weight.
+
+    Args:
+      index: integer indicates index of training weight to update.
+      tensor: tf.Tensor to update specific training weight.
+    """
+
+  @abc.abstractmethod
+  def compress_training_weights(self, *training_weights: tf.Tensor) -> List[tf.Tensor]:
+    """Define the operations to compress a single weight’s training form after training.
+
+    'compress_training_weights' can refer to making the weight more amenable to compression
+    or actually compress the weight.
+
+    The default is an identity.
+
+    Args:
+      *training_weights: tf.Tensors representing all variables used during
+        training, for a single compressible weight, in the order returned in
+        `init_training_weights`.
+
+    Returns:
+      List of tf.Tensors to set to compressed or more compressible form.
+    """
+
+  @abc.abstractmethod
+  def decompress_weights(self, *compressed_weights: tf.Tensor) -> tf.Tensor:
```

Comment:
This WeightCompressor only compress a single weight. (apply same logic for all compressible weights.)

For each compressible weight W (single tensor)
we compress a single tensor W to some list of compressed tensors.
So the result of decompress function is just a single tensor W.

e.g. weight clustering case, compressed tensors are two tensors. [lookup-table, indices]
W --(compress)--> [lookup-table, indices] --(decompress)--> W'
W` ~= W

Reply:
Slightly confused for a different reason:

In the case of weight clustering, if we want to compress kernel weights to cluster indices and centroids, we encounter a problem that the kernel weights are actually not differentiable in the wrapper layer training graph. How should we handle this case in the current compression APIs? Concretely,  In weight clustering APIs, we use a separate trainable variable `ori_weights_vars_tf` to copy the initial value of kernel attribute, then it acts as the gradient receiver during backprop since the original kernel attribute can be blocked by a `tf.argmin` op and can never be part of the training graph.

In the clustering use case [example](https://github.com/tensorflow/model-optimization/blob/c05ce9e35d6978579673c82e93e59ab475e8aa1e/tensorflow_model_optimization/python/core/common/keras/compression/algorithms/weight_clustering.py#L64), I can see the intention is to not update cluster indices? So in the training graph among original weights, indices and centroids, only the centroids are updated? This could work when the number of clusters is high (e.g. >=32) but it might give worse results for other cases.

Also, in the case of collaborative optimization technique clustering-preserving quantization aware training (CQAT), we do not have an option but to include cluster indices and `ori_weights_vars_tf` in the graph. Because doing otherwise will stop the min and max ranges learning from the original layer kernel weights.

In general, I think the compression APIs do not have to cover all use cases for now, and I am just trying to understand how to use them.

Reply:
Not xhark, but this use case can be handled by implementing your current training code in `project_training_weights`. The compress/decompress functions are intended to represent the graph model during "compression" and "inference" respectively and are not necessarily parts of the graph at training time. Init training weights will add trainable weights to the wrapper layer. You can treat these as any other layer weights and you can implement callbacks/losses as you need.